core/src/types/blocking_read/blocking_reader.rs

```diff
@@ -74,24 +84,26 @@ impl BlockingReader {
             }
         }
 
+        let iter = BufferIterator::new(
+            self.inner.clone(),
+            self.options.chunk(),
+            self.options.concurrent(),
+            start,
+            end,
+        );
+
         let mut bufs = Vec::new();
-        let mut offset = start;
-
-        loop {
-            // TODO: use service preferred io size instead.
-            let limit = end.map(|end| end - offset).unwrap_or(4 * 1024 * 1024) as usize;
-            let bs = self.inner.read_at(offset, limit)?;
-            let n = bs.remaining();
-            bufs.push(bs);
-            if n < limit {
-                return Ok(bufs.into_iter().flatten().collect());
-            }
 
-            offset += n as u64;
-            if Some(offset) == end {
-                return Ok(bufs.into_iter().flatten().collect());
+        for buffer in iter {
```

Comment:
Maybe we can use collect?

Reply:
How to return err early when using collect?

Reply:
> How to return err early when using collect?

collect will exit once error happened. See [`impl<A, E, V> FromIterator<Result<A, E>> for Result<V, E>`](https://doc.rust-lang.org/std/result/enum.Result.html#impl-FromIterator%3CResult%3CA,+E%3E%3E-for-Result%3CV,+E%3E):

Takes each element in the Iterator: if it is an Err, no further elements are taken, and the Err is returned. Should no Err occur, a container with the values of each Result is returned.